File: Binary Tree/101.对称二叉树.cpp
using namespace std;
/*
 * @lc app=leetcode.cn id=101 lang=cpp
 *
 * [101] 对称二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    bool compare(TreeNode *left, TreeNode *right)
    {
        if (left == nullptr && right == nullptr)
        {
            return true;
        }

        if (left != nullptr && right != nullptr)
        {
            if (left->val != right->val)
            {
                return false;
            }
            // 1左= 2右 ，1右 = 2左，同时相等才能条件成立
            return compare(left->left, right->right) && compare(left->right, right->left);
        }

        return false;
    }
    bool isSymmetric(TreeNode *root)
    {
        if (root == nullptr)
        {
            return true;
        }

        return compare(root->left, root->right);
    }
};
// @lc code=end

File: Binary Tree/二叉树的属性/101.对称二叉树.ts
/*
 * @lc app=leetcode.cn id=101 lang=typescript
 *
 * [101] 对称二叉树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isSymmetric(root: TreeNode | null): boolean {
  if (!root) {
    return false
  }

  let queue = [root.left, root.right]

  while (queue.length > 0) {
    // 每次取两个节点进行比对
    let left = queue.shift()
    let right = queue.shift()

    // 节点都不为空
    if (left && right) {
      // 判断值是否相等
      if (left.val !== right.val) {
        return false
      }
      // 将需要比较的节点按顺序继续放入队列
      queue.push(left.left)
      queue.push(right.right)
      queue.push(left.right)
      queue.push(right.left)
    } else if (!left && right) {
      return false
    } else if (left && !right) {
      return false
    }
  }
  // 遍历完成
  return true
};
// @lc code=end

